app/scripts/controllers/network/provider-api-tests/shared-tests.js

```diff
@@ -129,6 +138,377 @@ export function testsForRpcMethodAssumingNoBlockParam(method) {
+      // number is retrieved through the block tracker first. It doesn't
+      // matter what this is — it's just used as a cache key.
+      comms.mockNextBlockTrackerRequest();
+      comms.mockInfuraRpcCall({
+        request,
+        response: {
+          id: 12345,
+          jsonrpc: '2.0',
+          error: 'some error',
+          httpStatus: 420,
+        },
+      });
+      const promiseForResult = withInfuraClient(async ({ makeRpcCall }) =>
+        makeRpcCall(request),
+      );
+
+      await expect(promiseForResult).rejects.toThrow(
+        '{"id":12345,"jsonrpc":"2.0","error":"some error"}',
+      );
+    });
+  });
+
+  [503, 504].forEach((httpStatus) => {
+    it(`retries the request to Infura up to 5 times if it returns a ${httpStatus} response, returning the successful result if there is one on the 5th try`, async () => {
+      await withMockedInfuraCommunications(async (comms) => {
+        const request = { method };
+
+        // The first time a block-cacheable request is made, the latest block
+        // number is retrieved through the block tracker first. It doesn't
+        // matter what this is — it's just used as a cache key.
+        comms.mockNextBlockTrackerRequest();
+        // Here we have the request fail for the first 4 tries, then succeed
+        // on the 5th try.
+        comms.mockInfuraRpcCall({
+          request,
+          response: {
+            error: 'Some error',
+            httpStatus,
+          },
+          times: 4,
+        });
+        comms.mockInfuraRpcCall({
+          request,
+          response: {
+            result: 'the result',
+            httpStatus: 200,
+          },
+        });
+        const result = await withInfuraClient(
+          async ({ makeRpcCall, clock }) => {
+            // Note that we have to manually capture the value that the
+            // promise that `makeRpcCall` returns when it resolves or is
+            // rejected using `then`/`catch` instead of `try`/`catch` because
+            // the resolution/rejection occurs out of band. To be more
+            // specific, the promise will be fulfilled while the `while` loop
+            // below is running. This means that if we did not attach `catch`
+            // to the promise, Node would complain that the promise had a
+            // unhandled rejection. So we attach `catch` to capture the
+            // rejection and we also attach `then` to capture the resolution
+            // and check them after the loop ends.
+
+            let capturedResolutionValue;
+            let capturedRejectionValue;
+            makeRpcCall(request)
+              .then((resolutionValue) => {
+                capturedResolutionValue = resolutionValue;
+              })
+              .catch((rejectionValue) => {
+                capturedRejectionValue = rejectionValue;
+              });
+
+            while (nock.pendingMocks().length > 0) {
```

Comment:
This seems at risk of being an infinite loop, if these mocks remain pending. If we could rephrase this to avoid that possibility that would be ideal

Reply:
This was copypasta from other similar tests where we expect the promise to be rejected rather than resolved. I've removed this in [6440071](https://github.com/MetaMask/metamask-extension/pull/15717/commits/64400716b38bd792a090ff94e6b246d904d08816).

Reply:
Hmm, true. Nock doesn't seem to provide a way to get a list of mocked requests that have already been used, so I'm not sure how best to change this.

That said, the test timeout still applies here, so Jest should kill the loop along with the test if it goes past 2.5 seconds. Is that good enough?

Reply:
Not really - we'd still be punishing the user's CPU for that 2.5 seconds, and I don't trust Jest to be able to cleanly terminate a stuck process like that across all platforms (reliably terminating child processes is very difficult apparently, or so I've learned from being unable to find any tool that handles all edge cases).

Reply:
Maybe we could use a fixed number of retry attempts? e.g. if there are 5 requests we're expecting here, we could try 10 loops. Between waiting for a timer to be scheduled and running it, I wouldn't expect this to take more than 10 tries.

Reply:
I tried 10 iterations and that failed intermittently so I bumped it to 15: [4fc4b63](https://github.com/MetaMask/metamask-extension/pull/15717/commits/4fc4b63c3e3d3d9c4175c066a7eda44ccb41e9af)

Reply:
Looks good!

Though could we call the variable something a bit easier to understand the purpose of? e.g. `retries`

Reply:
Yes, good call.

By the way I noticed there were a few other cases that I missed. I also realized that there were other tests that handled this check in a slightly different way. Turns out we don't need to check that `nock.pendingMocks().length` is 0 because as long as an RPC call is busy retrying the request thanks to middleware, its promise won't get fulfilled. So we can check for promise fulfillment instead. I've made some refactors here: https://github.com/MetaMask/metamask-extension/pull/15717/commits/118271c9f62d3557c6834efdf06c175c855b2b34